security/acme-client/src/opnsense/mvc/app/library/OPNsense/AcmeClient/LeValidation/Base.php

```diff
@@ -0,0 +1,298 @@
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+ * AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+ * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+namespace OPNsense\AcmeClient\LeValidation;
+
+use OPNsense\Core\Config;
+use OPNsense\AcmeClient\LeAccount;
+use OPNsense\AcmeClient\LeUtils;
+
+/**
+ * LeValidation stub file, contains shared logic for all validation methods.
+ * @package OPNsense\AcmeClient
+ */
+abstract class Base extends \OPNsense\AcmeClient\LeCommon
+{
+    public const CONFIG_PATH = 'validations.validation';
+
+    /**
+     * The validation method cannot be properly initialized without the required
+     * configuration. LeValidation returns a more or less uninitialized object
+     * that first needs to be configured, and finally initialized by this function.
+     * @param $certid string the ID of the certificate object
+     * @param $accountuuid string the UUID of the account object
+     * @return bool
+     */
+    public function init(string $certid, string $accountuuid)
+    {
+        // Get config object
+        $this->loadConfig(self::CONFIG_PATH, $this->uuid);
+
+        // Get account object to query ID
+        $account = new LeAccount($accountuuid);
+        if (empty($account) || $account == null) {
+            LeUtils::log_error("unable to load account information: ${accountuuid}");
+            return false;
+        }
+
+        // Store auxiliary information (required to glue stuff together)
+        $this->cert_id = $certid;
+        $this->account_id = (string)$account->getId();
+        $this->account_uuid = (string)$account->getUuid();
+
+        // Teach acme.sh about DNS API hook location
+        $this->acme_env['_SCRIPT_HOME'] = '/usr/local/share/examples/acme.sh';
+
+        // Set log level
+        $this->setLoglevel();
+
+        // Set Let's Encrypt environment
+        $this->setEnvironment();
+
+        // Store acme hook
+        switch ((string)$this->config->method) {
+            case 'dns01':
+                $this->acme_args[] = '--dns ' . (string)$this->config->dns_service;
+                $this->acme_args[] = '--dnssleep ' . (string)$this->config->dns_sleep;
+                break;
+            case 'http01':
+                $this->acme_args[] = '--webroot /var/etc/acme-client/challenges';
+                break;
+        }
+
+        // Store acme filenames
+        $this->acme_args[] = '--home ' . self::ACME_HOME_DIR;
```

Comment:
@fraenki basically this either needs to use exec_safe() directly or you use 

$this->acme_args[] = '--home';
$this->acme_args[] = self::ACME_HOME_DIR;

and do the escapeshellargs on all the acme_args on command invoke...

Reply:
@fichtner Thanks for the explanation and the examples. :) Is https://github.com/opnsense/plugins/pull/2028/commits/9f16b9ea860fc5bfd36fb4775a2d45d16c91ea83 sufficient with this regard?

Reply:
Patch looks good but need to grep code base after merge to be sure. 👍